idaes/apps/grid_integration/multiperiod/price_taker_model.py

```diff
@@ -0,0 +1,1053 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+"""
+    Immediate ToDo list:
+        - Write methods to handle the representative day case
+            - automate cash flow expressions (likely use a 
+              weighted sum for representative days?)
+            - automate startup/shutdown constraints to
+              appropriately address the differences when
+              using representative days
+        - Add documentation to the IDAES readthedocs
+        - Add a warning when costs/revenues are given,
+          but are never used during cashflow construction
+"""
+import pandas as pd
+import numpy as np
+
+have_skl = True
+have_kn = True
+try:
+    from sklearn.cluster import KMeans
+except:
+    have_skl = False
+try:
+    from kneed import KneeLocator
```

Comment:
Are these existing dependencies, or are they adding new dependencies we need to be able to handle? If these are new dependencies, then this needs to be discussed in more detail (i.e. a meeting).

Reply:
@lbianchi-lbl FYI

Reply:
Linking to a previous discussion that explains the changes in setup.py: https://github.com/IDAES/idaes-pse/pull/1201#issuecomment-1585113725. If there is a better way to go about this, let us know.

Reply:
For optional dependencies, you can look into `attempt_import()` from `pyomo.common.dependencies` that streamlines the syntax a bit. You should be able to find usage  examples in the IDAES codebase and/or Pyomo docs.

Reply:
@lbianchi-lbl Can you provide a bit more context about optional dependencies? For this model, having access to a k-means clustering function is necessary, so what would be the purpose of an `attempt_import()`. In other words, if the import attempt failed, we would just have to display an error message saying something along the lines of "Install dependency XYZ". In which case, would there really be any difference between optional dependencies and actual dependencies?

Reply:
The purpose at `attempt_import` is for testing purposes to let us know if the tools we need are installed, and if not that we should skip these tests. The goal is to let us isolate potentially problematic dependencies and be able ot set up environments that do not have all dependencies without causing tests to fail.

Reply:
Alright, then scipy should suffice for now since the only clustering technique we've implemented so far is k-means clustering. But if we intend to add options for more clustering techniques in the future, then we'd likely need to reconsider using `attempt_import` or adding in the dependencies for sklearn and kneed